### ISSUE TITLE
Make custom loading indicators easier to implement

### ISSUE THREAD

**Issue by reinink** [inertiajs/inertia#288]
So, with addition of the new event system, it's now possible to use your custom loading indicators with Inertia. In fact, `@inertiajs/progress` itself uses the event system to do this ([see here](https://github.com/inertiajs/progress/blob/master/src/progress.js#L24-L26)).

My concern, however, is that implementing these custom loading indicators never turned out as simple as I had hoped it would. Here was my original vision for this:

```js
import NProgress from 'nprogress'
import { Inertia } from '@inertiajs/inertia'

Inertia.on('start', () => NProgress.start())
Inertia.on('progress', event => NProgress.set(event.detail.progress.percentage / 100 * .9))
Inertia.on('finish', () => NProgress.done())
```

(Note, this limits the progress to 90%, since 100% uploaded doesn't represent a finished visit, since you still need to wait for the server to respond).

And while this certainly works on the happy path, there are some issues and limitations to this approach:

## Preventing visits

If another `start` event listener calls `event.preventDefault()` (cancelling the visit), this implementation still starts the loading indicator, which is then never stopped, since the `finish` event never gets fired.

This can be solved by "waiting for the next tick", and then checking `event.defaultPrevented` on the event, like this:

```js
queueMicrotask(() => {
  if (event.defaultPrevented) {
    return
  }

  Inertia.on('start', () => NProgress.start())
})
```

And while this certainly works, the fact that you need this in order to properly implement your custom loading indicators is not easily discoverable.

## Cancelling visits

Even more difficult than handling prevented visits, is handling cancelled visits. More specifically, visits that are cancelled as a result of a visit to a new page (ex: clicking on another link before the current visit finishes).

The problem here is the order that the events get fired:

1. `start` is fired on the first page visit, which causes the progress indicator to start.
2. The user clicks another link, which causes a second `start` event to get fired (no change to the progress indicator).
3. The first visit asynchronously cancels, and fires the `finish` event, which causes the progress indicator to finish (clear).
4. The second visit continues to load, but now there is no progress indicator for it.
5. The second visit finishes, and fires the `finish` event, but there is no progress indicator to clear.

The heart of the issue here is the fact that the visit cancellation happens asynchronously, which means we can't force the `finish` event to fire before the new `start` event. (Ideally, I'd love to fix this, but I don't think if it's even technically possible, since I assume it's the browser that doing the `xhr` cancellation asynchronously.)

The solution here is to keep track of how many visits are currently in progress, and increment and decrement them as they start/finish, and then only clear the loading indicator when the count is back to zero.

You can see how I implement this in the `@inertiajs/progress` library using the `inProgress` variable ([see here](https://github.com/inertiajs/progress/blob/master/src/progress.js#L6)).

## Delay behaviour

The other thing the custom loading indicator implementation above doesn't handle is delayed showing of the loading indicator. When requests are fast (less than roughly 250ms), it's actually better to *not* show the loading indicator, since it really just causes a visual distraction to the page. Instead, it's better to only display the loading indicator in the event that the request is going slow. I have found 250ms to be a good threshold here. Turbolinks does the same thing, and actually delays for 500ms by default.

 To implement this properly requires some clever use of `setTimeout` and `clearTimeout`. You can see how I implement this in the `@inertiajs/progress` library using the `timeout ` variable ([see here](https://github.com/inertiajs/progress/blob/master/src/progress.js#L5)).

## Making things better

So, it's now clear to me that while the new event system can be used to implement your own custom loading indicator, in all likelihood, people are probably going to do it wrong, simply because it requires a pretty deep knowledge of how the event system works to get right. I think it would be better for Inertia to provide an abstraction layer over this that handles all this implementation detail for you.

This can be done in four ways:

## 1. A new progress package

The first option is to create a new generic progress package, similar to `@inertiajs/progress`, except without the NProgress integration. Under the hood this would use the event system, except it would provide three callbacks that can be used to implement your own custom loading indicator.

```js
import NProgress from 'nprogress'
import { progress } from '@inertiajs/generic-progress'

progress.init({
  delay: 500,
  onStart: () => NProgress.start(),
  onProgress: percentage => NProgress.set(percentage / 100 * .9),
  onFinish: () => NProgress.done(),
})
```

## 2. Add progress configuration to Inertia

The second option is to provide new configuration options into the Inertia adapters (and core). For example, we could add a new `loadingIndicator` prop that defines three callbacks:

```js
createApp({
  render: () => h(App, {
    initialPage: JSON.parse(el.dataset.page),
    resolveComponent: name => require(`./Pages/${name}`).default,
    loadingIndicator: {
    
      onStart: () => NProgress.start(),
      onProgress: percentage => NProgress.set(percentage / 100 * .9),
      onFinished: () => NProgress.done(),
    },
  }),
}).use(plugin).mount(el)
```

## 3. Add new progress events

The second option is a new set of events that we fire specifically for the purpose of the loading indicator:

```js
Inertia.on('progress-start', () => NProgress.start())
Inertia.on('progress-increment', event => NProgress.set(event.detail.percentage / 100 * .9))
Inertia.on('progress-finish', () => NProgress.done())
```

The nice thing about this approach is that progress packages can simply listen for these events, just like the `@inertiajs/progress`, without any dependencies on Inertia core.

What I really don't like is that we now have a `start` and `progress-start` event, a `progress` and `progress-increment` event, and a `finish` and `progress-finish` event. From for developer experience perspective, this seems really confusing. "What exactly is the difference between `start` and `progress-start`?"

I worry that this design just indicates a flaw in the event system design in general (see option four for more details).

## 4. Tweak the event system

I'd actually be interested in exploring what the implications of only firing the current `finish` event when <ins>all visits are done</ins> would be. Meaning, `finish` could technically get fired less times than `start`. I suspect this would actually be totally fine, since I suspect the primary use-case for the `finish` event is for hiding loading indicators. Further, the per-visit `onFinish` callback would still always be called, should you need to do clean up there.

Here is realistic NProgress implementation with this change...which honestly isn't bad.

```js
import NProgress from 'nprogress'
import { Inertia } from '@inertiajs/inertia'

let delay = 500
let timeout = null

Inertia.on('start', event => {
  queueMicrotask(() => {
    if (!event.defaultPrevented) {
      clearTimeout(timeout)
      timeout = setTimeout(() => {
        NProgress.set(0)
        NProgress.start()
      }, delay)
    }
  })
})

Inertia.on('progress', event => {
  if (NProgress.isStarted() && event.detail.progress.percentage) {
    NProgress.set(event.detail.progress.percentage / 100 * .9)
  }
})

Inertia.on('finish', () => {
  NProgress.done()
  clearTimeout(timeout)
})
```

This could be further simplified by adding a new new `before-start` event. This event would basically be exactly the same as `start`, except you'd only be able to cancel visits using `before-start`, and not via `start`. Meaning, `start` will only fire if `event.preventDefault()` wasn't called on the `before-start` event.

Here's an updated NProgress implementation for this change, which no longer requires the `queueMicrotask()` call and `event.defaultPrevented` check:

```js
import NProgress from 'nprogress'
import { Inertia } from '@inertiajs/inertia'

let delay = 500
let timeout = null

Inertia.on('start', () => {
  clearTimeout(timeout)
  timeout = setTimeout(() => {
    NProgress.set(0)
    NProgress.start()
  }, delay)
})

Inertia.on('progress', event => {
  if (NProgress.isStarted() && event.detail.progress.percentage) {
    NProgress.set(event.detail.progress.percentage / 100 * .9)
  }
})

Inertia.on('finish', () => {
  NProgress.done()
  clearTimeout(timeout)
})
```

As I write this, I'm really starting to think that the solution here is to change the `finish` event to only fire when <ins>all visits</ins> are done. This approach would allow us to avoid creating a new progress package, or adding additional configuration options to Inertia, or adding new progress specific events. Instead, we can simply document how to add custom loading indicators, sharing the above code snippet.

The only question left for me is whether it makes sense to add the new `before-start` event (which can be cancelled), and remove cancellation behaviour from the `start` event? My gut here is no, simply because there is a reasonable workaround using `queueMicrotask` and `event.preventDefault()`, and the amount of times you'll actually need to use this is going to be very minimal.

Basically this solution solves the most annoying/confusing part of displaying custom loading indicators, which is handling visit cancellations, where you get two `start` events fired and then two `finish` events fired.

**Comment by reinink** [comment]
It's worth noting that I actually had the global `finish` event firing only for the *currently active visit* at one time during the development of the event system:

https://github.com/inertiajs/inertia/blob/9ba22955be01846055edfe28f125318be66858c9/packages/inertia/src/inertia.js#L183-L187

However, I changed this, because it felt wrong to *always* call the `onFinish()` callback, but only fire the `finish` event for the active visit.

However, in a way this actually makes sense. When you're working with a specific visit request, it's important that the `onFinish()` callback is always called, since you may be doing cleanup work here. However, with the global events, there isn't really much added value in having it fire for both visits, especially considering that:

1. The `start` and `finish` events fire in the wrong order.
2. The `finish` event doesn't receive any visit/response data.

**Comment by reinink** [comment]
Before I make this change, I think I want to take another crack at solving the "Cancelling visits" event order issue. Speaking with @sebastiandedeyne and @claudiodekker, there might be an opportunity here to use a request manager to keep track of requests, which might help us with firing the events in the right order.

When making new visits:

1. Add a new `before-start` event that is fired when a new visit is initiated. This event is cancellable, which will prevent the visit from occurring.
2. We then immediately check if any existing visits are in progress. If they are, we trigger an `xhr` cancellation, and immediately fire the `finish` event.
3. We then fire a `start` event for the new visit.

This will cause the visit events to always be fired in the correct order:

1. User clicks link.
2. The `before-start` event is fired. If it is cancelled, the visit full stops.
3. The `start` event is fired.
4. While that request is still going, the user clicks another link.
5. The `before-start` event is fired. If it is cancelled, the first request just continues, and the new visit full stops.
6. The existing request is cancelled, and a `finish` event is fired.
7. The `start` event is fired for the new visit.
9. The `finish` event is fired when the new visit is done.

The only thing this doesn't solve is manual visit cancellations using `cancelToken.cancel()`. That happens outside of the flow of a new visit. I believe the solution here is to simply create our own `cancelToken` wrapper, that first calls `cancelToken.cancel()`, and then immediately fires the `finish` event.

However, in both of these cancellation situations, we're now firing the `finish` event early for these visits (to get the events firing in the right order), however, when the axios promise *actually finishes*, it's going to fire the `finish` event a second time for these visits. There is where a request manager is going to be useful. This will allow us to keep track of each visit, and only fire the necessary events once.
